### ISSUE TITLE
Angular2 sample requires second login attempt

### ISSUE THREAD

**Issue by Adtiv** [MicrosoftDX/kurvejs#37]
I've been playing around with this library in my angular2 app and got it working to where i can sign in to my MS email account. I been following this example: 
id.loginAsync().then(_ => {
    const graph = new kurve.Graph(id);
    graph.me.messages.GetMessages().then(messages => {
        console.log("my emails:");  
        messages.value.forEach(message => 
            console.log(message.subject)
        )
    });
});
I cant properly access anything from graph as they all return null, even me.getUser() although I can sign in and out correctly. I've added permissions to just about everything in the azure directory/graph for the project I set up, to see if I could get something but no luck. I also get a 400 (Bad Request) for the calendarView endpoint which is the one I need. Any suggestions as to how to fix this?

**Comment by billba** [comment]
Some things to confirm:
1. You can run the samples
2. You are using a valid Office 365 account associated with Exchange mail & calendar
3. You have correctly created an appId for either the V1 or V2 identity model
4. For a v1 app, you have correctly set permissions when registering the app

**Comment by Adtiv** [comment]
Thanks, I started with the samples and was having the same problem accessing data in the sample as I was in my application. I actually got GetUser() and events.GetEvents() to work only after removing the code that called calendarView and was giving me the 400(bad request). Here is some more info. on the calendarView error:  

"code": "ErrorInvalidParameter",
        "message": "This request requires a time window specified by the query string parameters StartDateTime and EndDateTime.",

Its not necessary I use calendarView since i can get what i need from the events endpoint, but thats the error I was getting anyways. Another issue I'm currently having is that it only accesses the events and user data after i run loginAsync twice. 

this is the code im using: 
` 
  loginMS() {
      this.id.loginAsync().then(_ => {
          console.log("id"+ this.id.isLoggedIn());
          this.isAuthenticated = true;
          console.log(this.isAuthenticated);
          const graph = new kurve.Graph(this.id);
          graph.me.GetUser().then(user=>{
            console.log(user.mail);
            this.msEmail=user.mail;
          })
          graph.me.events.GetEvents().then(events=>{
            console.log(events);
            events.value.forEach(event=>{
              event.title=event.subject;
              event.start=event.start.dateTime;
              event.end=event.end.dateTime;
              console.log(event.subject);
            })
            this.calendar.setMSevents(events.value);
          })
      });
  }`

it sais Im logged in and authenticated the first time i run loignMS but I get nothing for events and user.mail. The second time i run the function I get the correct data. I've tested this on a number of different browsers and the hosted site, It always requires two logins. Is there something else I'm missing?

**Comment by Adtiv** [comment]
Thanks, I started with the samples and was having the same problem accessing data in the sample as I was in my application. I actually got GetUser() and events.GetEvents() to work only after removing the code that called calendarView and was giving me the 400(bad request). Here is some more info. on the calendarView error:  

"code": "ErrorInvalidParameter",
        "message": "This request requires a time window specified by the query string parameters StartDateTime and EndDateTime.",

Its not necessary I use calendarView since i can get what i need from the events endpoint, but thats the error I was getting anyways. Another issue I'm currently having is that it only accesses the events and user data after i run loginAsync twice. 

this is the code im using: 

```
   loginMS() {
     this.id.loginAsync().then(_ => {
          console.log("logged In?"+ this.id.isLoggedIn());
          this.isAuthenticated = true;
          console.log(this.isAuthenticated);
          const graph = new kurve.Graph(this.id);
          graph.me.GetUser().then(user=>{
            console.log(user.mail);
            this.msEmail=user.mail;
          })
          graph.me.events.GetEvents().then(events=>{
            console.log(events);
            events.value.forEach(event=>{
              event.title=event.subject;
              event.start=event.start.dateTime;
              event.end=event.end.dateTime;
              console.log(event.subject);
            })
            this.calendar.setMSevents(events.value);
          })
      });
   }
```

it sais Im logged in and authenticated the first time i run loginMS but I get nothing for events and user.mail. The second time i run the function I get the correct data. I've tested this on a number of different browsers and the hosted site, It always requires atleast two logins. Is there something else I'm missing?

**Comment by billba** [comment]
I'm a little confused by your comments about calendarView. Where do we use it in the samples? FYI calendarView is quite a bit different than Events, especially when it comes to recurring appointments, which is why it exists as a separate endpoint. It requires two odata parameters. We provide a helper as noted in the QueryBuilder docs:

```
graph.me.calendarView.GetCalendarView(CalendarView.dateRange(start, end))
```

Your code looks okay to me, assuming that this.id was set up appropriately, e.g.

```
const id = new kurve.Identity("your-client-id", "http://your/path/to/login.html");
```

Does the VanillaJS sample also require two logins?

**Comment by Adtiv** [comment]
thanks, I never added that parameter in GetCalendarView(). I have just tried the angular2 sample. 
I am now seeing this error for the first login attempt: 

EXCEPTION: TypeError: _this.loginCallback is not a function
    at eval (identity.ts:189)
    at ZoneDelegate.invokeTask (zone.js:356)
    at Object.onInvokeTask (ng_zone_impl.ts:53)
    at ZoneDelegate.invokeTask (zone.js:355)
    at Zone.runTask (zone.js:256)
    at ZoneTask.invoke (zone.js:423)

second login attempt works fine and im connected to the api.

**Comment by billba** [comment]
Unfortunately the author of the Angular2 sample is on paternity leave, and I don't know anything about Angular2. Do the other samples work for you?

**Comment by Adtiv** [comment]
Appreciate the help anyway. I tested out the vanillajs sample and it didnt work for me using the sample application id. I tried running it in an http server and got the error that the reply address didnt work for the registered app. is there a certain port i should use? 

If the angular2 author comes back soon, heres some more insight on the issue Im having. 
If I follow the angular2 example exactly with the function running after pressing a login button:

```
    const id = new kurve.Identity(...)
    id.loginAsync().then(_ => {});
```

I still get "Uncaught TypeError: _this.loginCallback is not a function" for the first login attempt.
It redirects me, allows me to sign in, returns true for id.isLoggedIn() but doesnt give me any events or user information. All subsequent logins do not give me the TypeError but also dont give me events and user info. 

When i make id a class member and define it in the constructor and then reference it: 

```
    id; 
    constructor(){
        this.id= new kurve.Identity(...)
    }
    loginMS(){
       this.id.loginAsync().then(_ => {});
    }
```

I still get the TypeError and no data for the first login attempt, but the second one works. I also tried defining id in ngOnInit() with no luck.

**Comment by billba** [comment]
Yes, from the README:

> To run samples, start an http server using port 8000 in the root directory of this repository and aim your browser at http://localhost:port/samples.html

I'd love to know if you can run the VanillaJS sample!

**Comment by Adtiv** [comment]
Sorry bout that, vanillajs worked from port 8000!

**Comment by billba** [comment]
Okay then this is not a problem in Kurve, but instead something about the Angular2 sample. I'm going to retitle this issue and maybe Richard will surface from his new fatherhood and have a look at it :-)

**Comment by matvelloso** [comment]
Hi,

Kurve JS is just an experimental library. It's not an official library and it hasn't been updated for quite a while. I'd recommend instead using the new MSAL library which is our official one for this: https://github.com/AzureAD/microsoft-authentication-library-for-js
